### DISCUSSION TITLE
How to auto-update packages installed with pkg, pip?

### DISCUSSION THREAD

**Discussion by DominiqueSite** [termux/termux-packages#18938] [Q&A]
I have read [the wiki page](https://github.com/termux/termux-packages/wiki/Auto-updating-packages) regarding auto-updating packages in Termux, which left me with the following questions:
1. If I execute `pkg install python`, does the `python` package get automatically updated? The [build.sh](https://github.com/termux/termux-packages/blob/master/packages/python/build.sh) of `python` says `TERMUX_PKG_AUTO_UPDATE=false`, which I interpret as a no. If my interpretation is correct, what other ways are there to automatically update packages installed with `pkg`? I am specifically asking – and thankful – for a wiki page or line-by-line instructions.
2. Does this functionality also apply to python packages installed via `pip` (e.g. by executing `pip install radicale`)? If not, how do I enable automatic updates of python packages installed with `pip`? Again, I am specifically asking – and thankful – for a wiki page or line-by-line instructions.

**Comment by TomJo2000** [comment]
`TERMUX_PKG_AUTO_UPDATE` only applies to the building/rebuilding of packages on the repositories.
Packages for which it is enabled will be periodically checked for updates and the build system will attempt to automatically update them by incrementing to the new version number, updating the package checksums and starting a package build action.

This functionality is enabled for most "simple" packages.
Usually packages not requiring extensive patches.

It is explicitly disabled for "core" packages such as Python, which many other packages rely on.
And which require extensive compatibility testing before being updated.

---

This is mostly separate from updating packages on your device.
You can update your local packages using the `pkg` utility, which will wrap either `apt` (default) or `pacman`.

`pkg up` will perform a full update of all packages to the latest version available on the repositories of your selected Termux mirror.

Please be aware that Termux uses a "rolling release" model and **does not** support "partial upgrades".
Please update all of your packages when updating packages.

It is possible to script `pkg` to perform updates automatically, though this isn't recommended.

---

`pip` is provided as Python's language specific package manager and is offered without guarantee of compatibility.

Most pip modules should work just fine.
But some modules relying on natively compiled code, such as NumPy or Pandas **will not work on Termux without patches**, in such cases the modules may be offered as regular packages in the repositories NumPy and Pandas being `python-numpy` and `python-pandas` respectively.

`pip` could also be scripted to automatically update packages, though again this is not recommended and does not come with a guarantee of compatibility.

**Comment by sylirre** [comment]
`TERMUX_PKG_AUTO_UPDATE=false` is not relevant to Termux installations at all.

This variable is used ONLY by Termux developer scripts. When it is `true`, the script will automatically fetch new source version attempt to build it and if ok, push a package update to our Git repository.

Nothing to do with `pkg` or `pip`.

If you want automatic updates within Termux, then, please, write own script that would fulfill your requirements and launch it via `crond` or other automation methods such as Tasker + Termux:Tasker plug-in.

`yes | pkg upgrade -yq` - unconditionally upgrade all Termux packages except pip modules.

`pip3 install -y --upgrade module_1 module_2 ... module_N` - upgrade specific pip modules.

Pay attention that some modules may require different `pip` arguments.

Though, **Termux DOES NOT RECOMMEND to automatically upgrade packages unless you are prepared to get environment broken at some point.** *Yes, environment definitely will be broken with 100% probability when using automated updates.*

**Reply by evozi-juice** [reply]
I Used ChatGPT’s Advanced Voice Modeimport torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

model_name = "HuggingFaceH4/zephyr-7b-beta"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True, bnb_4bit_use_double_quant=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForCausalLM.from_pretrained(model_name, quantization_config=bnb_config)
tokenizer = AutoTokenizer.from_pretrained(model_name)pkg install pythonJames79x/autotrain-advanced
